csharp/src/Ice/OutputStream-Class.cs

```diff
@@ -67,92 +71,107 @@ public void IceEndSlice(bool lastSlice)
             RewriteByte((byte)_current.SliceFlags, _current.SliceFlagsPos);
         }
 
-        // Start writing a slice of a class or exception instance.
-        // This is an Ice-internal method marked public because it's called by the generated code.
-        // typeId is the type ID of this slice.
-        // firstSlice is true when writing the first (most derived) slice of an instance.
-        // slicedData is the preserved sliced-off slices, if any. Can only be provided when firstSlice is true.
-        // compactId is the compact type ID of this slice, if specified.
+        /// <summary>Starts writing the first slice of a class or exception instance. This is an Ice-internal method
+        /// marked public because it's called by the generated code.</summary>
+        /// <param name="allTypeIds">The type IDs of all slices of the instance (excluding sliced-off slices), from
+        /// most derived to least derived.</param>
+        /// <param name="slicedData">The preserved sliced-off slices, if any.</param>
+        /// <param name="errorMessage">The exception error message (provided only by exceptions).</param>
+        /// <param name="compactId">The compact ID of this slice, if any. Used by the 1.1 encoding.</param>
         [EditorBrowsable(EditorBrowsableState.Never)]
-        public void IceStartSlice(string typeId, bool firstSlice, SlicedData? slicedData = null, int? compactId = null)
+        public void IceStartFirstSlice(
+            string[] allTypeIds,
+            SlicedData? slicedData = null,
+            string? errorMessage = null,
+            int? compactId = null)
         {
-            Debug.Assert(InEncapsulation && _current != null);
-            if (slicedData.HasValue)
+            Debug.Assert(InEncapsulation && _current.InstanceType != InstanceType.None);
+
+            if (slicedData is SlicedData slicedDataValue)
             {
-                Debug.Assert(firstSlice);
+                bool firstSliceWritten = false;
```

Comment:
I don't see why this bool is useful. Can't we just call IceStartNextSlice in the `try {}` block after `WriteSlicedData`?

Reply:
I added this bool to make it clear that only `WriteSlicedData` can throw NotSupportedException (which we ignore and then continue).

If we remove the bool and call `IceStartNextSlice` right after `WriteSlicedData`, it would sound like we'd ignore/continue on `NotSupportedException` thrown by `IceStartNextSlice`.